Feature: macOS, watchOS, tvOS support

Brett-Best opened the pull request.

Commit:
Add callbacks for after and every to account for cases where you want to get timer for after and don't want to use it for every

Commit:
Update SFSymbolName to v2 including all appropriate deprecations and availability annotations for backwards compatibility

Commit:
Squashed commit of the following:

commit 9bfac36c7d6560252260c82662f83744c86e6f3f
Author: Adam Fordyce <adam@pureswiftui.com>
Date:   Sun Nov 1 22:41:01 2020 +0000

    Add convenience initialisers to Color

commit 36bf741ef360144c96f4750eb5017ab183ab24a5
Author: Adam Fordyce <adam@pureswiftui.com>
Date:   Sun Oct 25 22:17:00 2020 +0000

    Remove @inlinable - no longer needed to address segmentation fault introduced in Xcode 11.4 - fixed in 11.5

commit 252427d33015cfc689e29f09164cad92cd351811
Author: Adam Fordyce <adam@pureswiftui.com>
Date:   Sun Oct 25 21:31:20 2020 +0000

    Add convenience initialisers for Gradients

    Add xy scaling for layout guides
    Add convenience initialisers for CGRect
    Add "to" extension to CG types
    Add "sizeScaled" to GeometryProxy
    Add "circle" to Path

Commit:
Add xScale/yScale operators to layout guides and configs and include tests

Commit:
Remove UINumericType for faster compilation times and remove deprecated code

Commit:
Add initialiser on SFSymbol that takes a raw String and remove opacity modifier causing slow builds

Commit:
Merge branch 'feature/2.0.0' of https://github.com/CodeSlicing/pure-swift-ui into feature/2.0.0

# Conflicts:
#	Sources/PureSwiftUI/Extensions/SwiftUI/Views/Modifiers/View+Modifiers.swift

Commit:
Remove slow performing modifiers and further adopt CGFloat as main type in layout guides

Commit:
Clean up redundant modifiers adding default values and migrate preview device names

Commit:
Remove argument label from inset call

Commit:
Rename relativeArc in Path extension

Commit:
Update podspec and readme for v2.0.0

Commit:
Add static initialiser for Color and update readme and podspec for v2.0.1

Commit:
Add circle overloads to Path for various parameters

Commit:
Add extensions for drawing circles on Path

Brett-Best committed:
Added support for macOS, tvOS, watchOS

Brett-Best commented on Sources/PureSwiftUI/Extensions/Convenience/CoreGraphics/CGRect+Convenience.swift:
```diff
@@ -183,9 +183,13 @@ public extension CGRect {
-    
+
     func inset(_ top: CGFloat, _ leading: CGFloat, _ bottom: CGFloat, _ trailing: CGFloat) -> CGRect {
-        inset(by: UIEdgeInsets(top: top, left: leading, bottom: bottom, right: trailing))
+      #if os(macOS)
+      return self
+      #else
+      return inset(by: UIEdgeInsets(top: top, left: leading, bottom: bottom, right: trailing))
+      #endif
     }
```
I wasn’t sure what to replace this with on macOS, would be great if you could take a look!

Brett-Best commented on Sources/PureSwiftUI/ViewModifiers/Font/DynamicallyScaledFontViewModifier.swift:
```diff
@@ -14,7 +14,11 @@ public struct DynamicallyScalingFontViewModifier: ViewModifier {
     var weight: Font.Weight?
 
     public func body(content: Content) -> some View {
+        #if os(macOS)
+        let scaledSize = CGFloat(17)
+        #else
         let scaledSize = UIFontMetrics.default.scaledValue(for: size)
+        #endif
         return content.font(createFont(scaledSize: scaledSize))
     }
```
I wasn’t sure what to replace this with either on macOS, I wonder what the macCatalyst implementation calls through to in AppKit. Would be great if you could take a look, perhaps having a hardcoded value is okay.

OskarGroth replied:
Dynamic type does not exist on macOS. The docs for the `sizeCategory` environment value (now renamed `dynamicTypeSize`) states that:

> On macOS, this value cannot be changed by users and does not affect the text size

Just make this whole modifier unavailable on macOS.

CodeSlicing replied:
yes completely agree. Once this PR is applied, I'll move the conditional compilation to include the entire extension.

Brett-Best reviewed (commented):
@CodeSlicing would appreciate if you could have a look at this!
I could also make a PR for making sure the project builds (+ maybe run the unit tests), see https://github.com/SwiftUIX/SwiftUIX/runs/1471421259 for an example of what I’ve setup (:

Commit:
Update control points so they work with strokes and fills when part of a closed subpath

Commit:
Update readme and podspec for v2.0.3

Commit:
Remove old control point logic

Commit:
Updated control point rendering for quad curves to match curve styling

Brett-Best committed:
Merge branch 'feature/2.0.0' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

Commit:
Add convenience properties and functions to GeometryProxy, and UIScreen

Remove conflict with native scaling for Shape
Add convenience initialisers for Label and SFSymbols
Add renderingMode to SFSymbol constructor functions

Commit:
Resolve ambiguity with native offset modifier on shape

Brett-Best committed:
Merge branch 'feature/2.0.0' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

# Conflicts:
#	Sources/PureSwiftUI/GlobalFunctions/ConstructorFunctions/SFSymbol.swift

Commit:
Fix shadowIfNot when using explicit y offset

Expand signature of static initialisers on color to take color space into account as well as opacity. Add static initialiser for white

Brett-Best committed:
Merge branch 'feature/2.0.0' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

Commit:
Add static initialisers and convenience functions to CGRect and GeometryProxy

Add tests to cover new functionality in CGRect

CodeSlicing commented:
Hey there,

Sorry for the lack of progress on this! I have so little time for anything at the moment that it's difficult to carve out the time to do a proper review.

V interested in the automated tests on commit.

Promise to get back to you on this as soon as I can dedicate the time to it. 

All the best,

Adam

CodeSlicing commented:
Once I've got everything merged back to develop and looking good as it stands, I'll make a move on this your PR :)

Brett-Best committed:
Merge branch 'feature/2.0.0' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

Commit:
Add support for relative and floating point coordinates

Commit:
Add relative coordinate functionality to layout guides

Rename xScaled and yScaled to relativeX and relativeY in CGRect
Add convenience subscript to CGRect to extract relative coordinates
Add convenience "to" function to Angle
Add pre-built gradient convenience initialisers for all gradient types
Add tests covering new layout guide functionality

Commit:
Add version info for 2.1.0 to readme

Brett-Best committed:
Merge branch 'feature/2.0.0' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

Brett-Best committed:
Merge branch 'develop' of github.com:CodeSlicing/pure-swift-ui into feature/MPP-Support

# Conflicts:
#	README.md
#	Sources/PureSwiftUI/Extensions/Convenience/CoreGraphics/CGRect+Convenience.swift
#	Sources/PureSwiftUI/Extensions/SwiftUI/Shapes/Shape+Modifiers.swift
#	Sources/PureSwiftUI/GlobalFunctions/ConstructorFunctions/SFSymbol.swift

Brett-Best commented:
@CodeSlicing be great if you could take a look at this before WWDC come around, was a bit of a pain to merge your latest changes in

CodeSlicing commented:
Sorry about that - had to merge everything to develop to get things back on track. 

As it stands the plan is to separate the drawing and many of the lower-level convenience extensions into a separate package that will then be imported by PureSwiftUI. It's something people have been asking for and makes sense to me. 

With that in mind, I've decided it's version 3 of PureSwiftUI that's going to get your changes to match up with the next version of SwiftUI. I'll be taking time off work to specifically look at things like this. 

I realise it's frustrating. This is why in the readme I was quite explicit in not open-sourcing this project as I have so little time to administer PRs and I don't want to mess people about. However, I do have a plan to merge the changes you've been kind enough to work on :)

Brett-Best commented:
No need to apologise 😊, that sounds great to me. Fingers crossed Xcode 13 won’t require many changes to get things compiling!

OskarGroth commented:
@CodeSlicing I'd love to see macOS support for this lib. Any reason why this PR can't be merged as-is? I get that you're planning for major refactors but it seems to be pretty minor changes and life tends to get in the way (we're soon 1 year down from this PR).

CodeSlicing commented:
Yes indeed, I'm currently working on 3.0.0 of PureSwiftUI which will incorporate this PR! The idea is to release 3.0.0 in September following the official release of iOS15 etc.

CodeSlicing commented:
Ok merging this into feature/2.3.0 and will cherry pick into feature/3.0.0 of PureSwiftUI and feature/1.0.0 of PureSwiftUIDesign.

CodeSlicing commented:
PR is now part of tag '2.3.0-beta-2' - will update this PR when appropriate changes have been cherry picked into the other versions I mentioned.

CodeSlicing commented:
scratch that - it's now '2.3.0-beta-4' - had to make some changes in the conditional compilation for watchOS

CodeSlicing commented:
All very exciting stuff! Thanks Brett, and sorry it took sooooooo long ;o)
